Samples/WindowPlacement/cpp/inc/CurrentMonitorTopology.h

```diff
@@ -0,0 +1,726 @@
+    // with 'force' parameter. This forces us to reload the monitor data
+    // at this awareness.
+    if (!m_validDpiAwarenesses[awareness])
+    {
+        m_validDpiAwarenesses[awareness] = true;
+
+        RefreshMonitorData(true /* force */);
+    }
+
+    // Call the function for each monitor in the list.
+    for (const auto& monitor : m_monitorLists[awareness])
+    {
+        if (!fn(monitor))
+        {
+            break;
+        }
+    }
+}
+
+/* static */
+inline BOOL
+CurrentMonitorTopology::MonitorEnumProc(
+    HMONITOR handle,
+    HDC,
+    PRECT,
+    LPARAM pMonListPtr)
+{
+    // Get the data for this monitor.
+    // This can fail (HMONITORs can become invalid at any time).
+    MonitorData monData;
+    if (!MonitorData::FromHandle(handle, &monData))
+    {
+        return false;
+    }
+
+    // Allocate a SharedMonitorPtr (std::shared_ptr<const MonitorData>).
+    SharedMonitorPtr monitorData = SharedMonitorPtr(new MonitorData(monData));
+    if (!monitorData)
+    {
+        return false;
+    }
+
+    // Add the shared monitor pointer to the list provided by the caller.
+    auto pMonitorList = reinterpret_cast<std::vector<SharedMonitorPtr>*>(pMonListPtr);
+    pMonitorList->push_back(monitorData);
+
+    return true;
+}
+
+inline void
+CurrentMonitorTopology::RefreshMonitorData(bool force)
+{
+    // There is a new API GetCurrentMonitorTopologyId, added in GE_RELEASE.
+    //
+    // We use this API if it is available to know the current topology ID.
+    // This makes this ID global (shared with all apps).
+    //
+    // Using the system's ID also allows us to skip unneeded work when we get
+    // several WM_DISPLAYCHANGE in a row. It's likely we receive the first one
+    // after all changes to the monitors have been made, so we can throw out
+    // subsequent messages by comparing the topology ID and finding it hasn't
+    // changed.
+    const UINT topologyId = GetCurrentMonitorTopologyId();
+    if (topologyId)
+    {
+        // If called with force, we're being called because the thread changed
+        // its DPI awareness, requiring that we rebuild the monitor lists
+        // because the number of awarenesses we're maintaining has changed, not
+        // because we think the monitors have changed.
+        if (!force && (m_lastTopologyId == topologyId))
+        {
+            return;
+        }
+        m_lastTopologyId = topologyId;
+    }
+    else
+    {
+        // Fallback (topology ID API not available): Increment the ID every
+        // time we receive a display change or setting change.
+        // This ID could be used by someone using this class in the same way
+        // that we use the API when it is available (to know if anything about
+        // the monitors may have changed between two points in time).
+        m_lastTopologyId++;
+    }
+
+    bool succeeded = true;
+
+    m_numMonitors = GetSystemMetrics(SM_CMONITORS);
```

Comment:
Is this a race condition? Can't the number of monitors change at any point after this?